core/shared/src/main/scala/zio/stm/ZSTM.scala

```diff
@@ -1632,282 +1665,265 @@ object ZSTM {
+       */
+      private[internal] def commit(): Unit = {
+        val it = map.valuesIterator
+        while (it.hasNext) it.next.commit()
+      }
 
-    /**
-     * Atomically collects and clears all the todos from any `TRef` that
-     * participated in the transaction.
-     */
-    def collectTodos(journal: Journal): MutableMap[TxnId, Todo] = {
-      import collection.JavaConverters._
+      /**
+       * Collects all todos and submits them to the executor.
+       *
+       * '''NOTE''': This method MUST be invoked while we hold the lock on the
+       * journal
+       */
+      private[internal] def completeTodos(executor: Executor)(implicit unsafe: Unsafe): Unit = {
+        val todos = collectTodos()
+        if (todos.nonEmpty) executor.submitOrThrow(() => execTodos(todos))
+      }
 
-      val allTodos  = new MutableMap[TxnId, Todo](DefaultJournalSize)
-      val emptyTodo = Map.empty[TxnId, Todo]
+      /**
+       * Executes the todos in the current thread, sequentially.
+       */
+      private[this] def execTodos(todos: Map[TxnId, Todo]): Unit = {
+        val it = todos.valuesIterator
+        while (it.hasNext) it.next.apply()
+      }
 
-      val it = journal.entrySet.iterator
-      while (it.hasNext) {
-        val tref = it.next.getValue.tref
-        val todo = tref.todo
+      /**
+       * Flag indicating whether the journal is invalid
+       *
+       * '''NOTE''': This method MUST be invoked while we hold the lock on the
+       * journal
+       */
+      private[ZSTM] def isInvalid: Boolean = !isValid
 
-        var loop = true
-        while (loop) {
-          val oldTodo = todo.get
+      /**
+       * Flag indicating whether the journal is valid
+       *
+       * '''NOTE''': This method MUST be invoked while we hold the lock on the
+       * journal
+       */
+      private[ZSTM] def isValid: Boolean = {
+        val it = map.valuesIterator
+        while (it.hasNext) if (!it.next().isValid) return false
+        true
+      }
 
-          loop = !todo.compareAndSet(oldTodo, emptyTodo)
+      /**
+       * Creates a function that can reset the journal.
+       */
+      private[ZSTM] def resetFn(): () => Unit = {
+        val currentNewValues = ZSTMUtils.newMutableMap[TRef[_], Any](map.size)
+        val itCapture        = map.iterator
+        while (itCapture.hasNext) {
+          val (key, value) = itCapture.next()
+          currentNewValues.update(key, value.unsafeGet[Any])
+        }
 
-          if (!loop) allTodos.putAll(oldTodo.asJava): @nowarn("msg=JavaConverters")
+        () => {
+          val saved = ZSTMUtils.newMutableMap[TRef[_], Entry](map.size)
+          val it    = map.iterator
+          while (it.hasNext) {
+            val (key, value) = it.next()
+            val resetValue = currentNewValues.getOrElse(key, null) match {
+              case null  => value.expected
+              case value => value
+            }
+            saved.update(key, value.copy().reset(resetValue))
```

Comment:
inline this into the pattern match above, it seems you coud avoid ony copy/reset in the `null` branch?

Reply:
It seems that we _might_ be able to avoid the `copy()` but not the `reset` (tests are failing if I don't reset in the `null` case).

However, this method is only used for `ZSTM#orElse` so I'd like to play it safe here even if slightly less performant just in case there's an edge case that we're not capturing in the test suite. WDUT?

Reply:
sure, I think if the reset is necessary then the copy definitely is too